### PR TITLE
Restructuring to acommodate new features

### LINKED ISSUE #20: Add total hours of sun and remaining sun hours
Add to the card 2 new elements to display the total hours of sun in the day and the remaining hours.
More context: https://www.reddit.com/r/homeassistant/comments/ne2mx4/sun_card_for_home_assistant/gyh7yrs

### LINKED ISSUE #22: Allow to hide each text element
Keeping the current visibility as default, we could allow each text element to be hidden.

Context: https://www.reddit.com/r/homeassistant/comments/ne2mx4/sun_card_for_home_assistant/gyeip4x

### PR DESCRIPTION
**Pull request by bjornomy**
This PR contains a major rewrite/restructure of this component. This is by no means trying to undermine the **great** work of the creator.

The goal was to make it easier to develop new features and accommodate for more configuration, as well as introducing some opinionated best practices.

An (non-exhausted) list of changes/additions:
* extract components to separate files
  * rewrite `render` pipeline with composition using extracted components (#22)
  * use new configuration options in composition
* tweak how we read config
* add more configuration options
    * any component other than the graph can now be hidden through the `fields` configuration option (#22)
    * which sun component to use is now configurable (#23)
* use full Date in all positional logic
* streamline internationalization
  * use lowercase translation keys
  * implemented arbitrary nesting
  * implement date formatting using `Intl.DateTimeFormatter` (NOTE: one feature here requires es2020)
  * implement I18N class to handle all internationalization
* extract some common functionality to `utils.ts`
* try to unify css / simplify (#19)
* extract graph colors to css

### CODE REVIEW

**Comment by AitorDB** [comment]
Awesome! The list of changes/additions looks really good. It may take me a few days until I can properly take a look to this because of lack of time, but looks promising. Thanks!

File: src/cardStyles.ts
Comment (AitorDB) [inline src/cardStyles.ts]: Typo, finxing it should also fix the issue with the dusk gradient

File: src/cardStyles.ts
Comment (AitorDB) [inline src/cardStyles.ts]: Lines are not too visible on light mode, I would say to change it to something like `#d7d7d7` maybe?

File: src/utils.ts
Comment (AitorDB) [inline src/utils.ts]: Awesome! Way more organized having utils! I was thinking, could we make utils a folder and move internationalization to its own file? I will be adding more feature soon an I think it would be better to split it before it gets too big

File: src/utils.ts
Comment (AitorDB) [inline src/utils.ts]: Could we rename this to translate or similar? same with the parameter name

File: src/utils.ts
Comment (AitorDB) [inline src/utils.ts]: Could we transform this into a export class with static methods instead?

File: README.md
Comment (AitorDB) [inline README.md]: I think this should be taken by default from the language and only if specified then enforce one mode or the other

File: src/index.ts
Comment (AitorDB) [inline src/index.ts]: I think this is missing a assignation, changing language in config is not changing the language of the card

File: src/index.ts
Comment (AitorDB) [inline src/index.ts]: Could we trigger the card error instead for this? In that way people should be able to see the problem easily when trying to set a language that is not supported. Also, I would change the copy to `... is not a supported language` instead.

File: src/components/SunErrorContent.ts
Comment (AitorDB) [inline src/components/SunErrorContent.ts]: Can we change to single quotes here and in index?

File: src/components/SunCardContent.ts
Comment (AitorDB) [inline src/components/SunCardContent.ts]: Can we change this to have 2 spaces?

File: src/utils.ts
Comment (AitorDB) [inline src/utils.ts]: There's a space on `mozilla` that breaks the link

**Review by AitorDB — CHANGES_REQUESTED** [review]
Awesome work! There're a few things but it looks great! Happy to discuss any of the comments I left! Sorry it took me some time to check the PR

File: src/cardStyles.ts
Comment (bjornomy) [reply src/cardStyles.ts]: Correcting typo `Inital` -> `Initial` for all relevant css-classes

File: src/cardStyles.ts
Comment (bjornomy) [reply src/cardStyles.ts]: I tried to reuse the colors, since I'm not god at that stuff, and hoped you would suggest better ones 😄. Changed to `#d7d7d7`

File: src/utils.ts
Comment (bjornomy) [reply src/utils.ts]: That is fine. Created folder `utils` containing `I18N.ts` and `HelperFunctions.ts`.

I'm not that well versed in TypeScript, and though putting everything in it's own class was unnecessary for the helper functions

File: src/utils.ts
Comment (bjornomy) [reply src/utils.ts]: I was kind of following Vue's example. My main reason to keep `I18N.tr(...)` is the short and easy method name, who most should understand. Keeps the code from getting 'cluttered' with `I18N.translate(...)` all over the place, and you might loose focus on what we translate, which is the important part. I'm not opposed, but hesitant 😆 .

I'll change the parameter.

Comment (bjornomy) [reply src/utils.ts]: Added some 'Javadoc' style comments to the method which might help

File: src/utils.ts
Comment (bjornomy) [reply src/utils.ts]: Done: `utils/HelperFunctions.ts`

File: src/index.ts
Comment (bjornomy) [reply src/index.ts]: Good spot! Fixed.

File: src/index.ts
Comment (bjornomy) [reply src/index.ts]: I was trying to make use of a feature mentioned in the [Lovelace docs](https://developers.home-assistant.io/docs/frontend/custom-ui/lovelace-custom-card/):

> Home Assistant will call setConfig(config) when the configuration changes (rare). If you throw an exception if the configuration is invalid, Lovelace will render an error card to notify the user.

If we want to use our custom error card, we can possibly do that as well.

Changed `[V|v]alid` -> `[S|s]upported`

File: src/components/SunErrorContent.ts
Comment (bjornomy) [reply src/components/SunErrorContent.ts]: Could you perhaps provide some more details? I don't get/understand the issue.

File: src/components/SunCardContent.ts
Comment (bjornomy) [reply src/components/SunCardContent.ts]: Fixed

File: src/utils.ts
Comment (bjornomy) [reply src/utils.ts]: Fixed

File: README.md
Comment (bjornomy) [reply README.md]: I was hoping `Intl.DateTimeFormatter` would handle this for me, even with the flag present in the options, but it seems my assumption was wrong.

I'll take a look and try to implement the flag with `undefined` to support this behavior

File: src/index.ts
Comment (bjornomy) [reply src/index.ts]: Note that `hass.locale.language` overrides `hass.language`

**Comment by bjornomy** [comment]
Thanks for good input! I don't particularly mind the wait, as long as we don't merge too much stuff into `develop`, which might cause conflicts.
I'll rebase when we have resolved all conversations.

File: src/components/SunErrorContent.ts
Comment (bjornomy) [reply src/components/SunErrorContent.ts]: Nvm, I got it! Changed all double quote imports I could find to single quote.

File: README.md
Comment (bjornomy) [reply README.md]: Updated to to defer from locale when not set in configuration.

File: src/utils.ts
Comment (AitorDB) [reply src/utils.ts]: Yes, it's not necessary but it helps to keep some consistency, it could be consider some a bit subjective but if there's not a reason to not do it, I'll rather go in that direction

File: src/utils.ts
Comment (AitorDB) [reply src/utils.ts]: Yes, I think the doc comment should be enough. Being used also by Vue i18n and maybe others we could argue that is more standard too, so I'm ok with it :D

File: src/index.ts
Comment (AitorDB) [reply src/index.ts]: Cool! I wasn't aware of that, I'm ok with throwing the error then, it's cleaner definitely

File: src/index.ts
Comment (AitorDB) [reply src/index.ts]: Yes that seems right to me, I would expect `hass.language` to be removed at some point (or maybe not if they decide to keep it for retrocompatibility) although I think we will have to keep it (not a big issue in the card code imo)

File: src/types/index.ts
Comment (AitorDB) [inline src/types/index.ts]: I think you meant `ESunCardI18NKeys` instead `ESunCardErrors`?

Comment (bjornomy) [reply src/types/index.ts]: Yes, my bad. Fixed

**Comment by bjornomy** [comment]
The branch is now rebased on the latest `develop`

**Comment by AitorDB** [comment]
I just built it and ran some tests on Home Assistant, it needs a small fixes on Header and Footer because `this.times` and `this.data` can be `undefined` sometimes. An easy fix would be to just change `this.fields?.dawn` to `this.fields?.dawn && this.times?.dawn` and similar with the rest of elements, but I'll leave that to your choise. Also we will have to tweak styles a little bit, but happy to do that myself if you don't want to or as part of other PR before releasing.

PC size

Phone size

**Comment by bjornomy** [comment]
> I just built it and ran some tests on Home Assistant, it needs a small fixes on Header and Footer because `this.times` and `this.data` can be `undefined` sometimes. An easy fix would be to just change `this.fields?.dawn` to `this.fields?.dawn && this.times?.dawn` and similar with the rest of elements, but I'll leave that to your choise. Also we will have to tweak styles a little bit, but happy to do that myself if you don't want to or as part of other PR before releasing.
> 
> PC size
> 
> Phone size

Tried to fix this in the helper function. Please tell me if I misunderstood.

As for styles, I think it would be best to do it in a separate issue, as this one is a bit large already 😅

**Comment by AitorDB** [comment]
> > I just built it and ran some tests on Home Assistant, it needs a small fixes on Header and Footer because `this.times` and `this.data` can be `undefined` sometimes. An easy fix would be to just change `this.fields?.dawn` to `this.fields?.dawn && this.times?.dawn` and similar with the rest of elements, but I'll leave that to your choise. Also we will have to tweak styles a little bit, but happy to do that myself if you don't want to or as part of other PR before releasing.
> > PC size
> > Phone size
> 
> Tried to fix this in the helper function. Please tell me if I misunderstood.
> 
> As for styles, I think it would be best to do it in a separate issue, as this one is a bit large already

Yes, sorry, I didn't add enough information, the issue is for example on [`SunCardHeader.ts:L36`](https://github.com/AitorDB/home-assistant-sun-card/pull/64/files#diff-e28b4807eef5e5d40d8185148a06429ce0448a45858baf5ab3b55e958000167bR36), the code try to access `this.time.sunrise`, however `this.times` sometimes can be undefined, which throws an error `Cannot read property 'sunrise' of undefined`.

Happy to keep styles for another moment

**Comment by bjornomy** [comment]
Ahh, that was an oversight on my part. Should be fixed now.

**Review by AitorDB — APPROVED** [review]
Working now! Thank you so much! :smile:

**Pull request merged**
